src/encryption.rs

```diff
@@ -0,0 +1,357 @@
+        }
+
+        if new_annotations.is_empty() {
+            return Err(anyhow!("no encryptor found to handle encryption"));
+        }
+
+        new_annotations.insert(
+            "org.opencontainers.image.enc.pubopts".to_string(),
+            base64::encode(pub_opts),
+        );
+
+        Ok(new_annotations)
+    }
+}
+
+/// get_key_wrapper looks up the encryptor interface given an encryption scheme (gpg, jwe)
+pub fn get_key_wrapper(scheme: &str) -> Result<&dyn KeyWrapper> {
+    match KEY_WRAPPERS.get(scheme) {
+        Some(key_wrapper) => Ok(key_wrapper),
+        None => Err(anyhow!("key wrapper not supported!")),
+    }
+}
+
+/// get_wrapped_keys_map returns a option contains map of wrapped_keys
+/// as values and the encryption scheme(s) as the key(s)
+pub fn get_wrapped_keys_map(desc: OciDescriptor) -> Option<HashMap<String, String>> {
+    desc.annotations.as_ref()?;
+
+    let mut wrapped_keys_map = HashMap::new();
+    for (annotations_id, scheme) in KEY_WRAPPERS_ANNOTATIONS.iter() {
+        if let Some(annotations) = desc.annotations.as_ref() {
+            if let Some(anno_id) = annotations.get(annotations_id) {
+                wrapped_keys_map.insert(scheme.clone(), anno_id.clone());
+            }
+        }
+    }
+
+    Some(wrapped_keys_map)
+}
+
+// pre_wrap_keys calls wrap_keys and handles the base64 encoding and
+// concatenation of the annotation data.
+fn pre_wrap_key(
+    keywrapper: &dyn KeyWrapper,
+    ec: &EncryptConfig,
+    mut b64_annotations: String,
+    opts_data: &[u8],
+) -> Result<String> {
+    let new_annotation = keywrapper.wrap_keys(ec, opts_data)?;
+    if new_annotation.is_empty() {
+        return Err(anyhow!("new annotations is empty!"));
+    }
+
+    let b64_new_annotation = base64::encode(new_annotation);
+    if b64_annotations.is_empty() {
+        return Ok(b64_new_annotation);
+    }
+
+    b64_annotations.push(',');
+    b64_annotations.push_str(&b64_new_annotation);
+    Ok(b64_annotations)
+}
+
+// pre_unwrap_key decodes the comma separated base64 strings and calls the unwrap_key function
+// of the given keywrapper with it and returns the result in case the unwrap_key functions
+// does not return an error. If all attempts fail, an error is returned.
+fn pre_unwrap_key(
+    keywrapper: &dyn KeyWrapper,
+    dc: &DecryptConfig,
+    b64_annotations: &str,
+) -> Result<Vec<u8>> {
+    if b64_annotations.is_empty() {
+        return Err(anyhow!("annotations is empty!"));
+    }
+
+    let mut errs = String::new();
+    for b64_annotation in b64_annotations.split(',') {
+        let annotation = base64::decode(b64_annotation)?;
+
+        match keywrapper.unwrap_keys(dc, &annotation) {
+            Err(e) => {
+                errs.push_str(&e.to_string());
+                continue;
+            }
+            Ok(opts_data) => {
+                return Ok(opts_data);
+            }
+        };
+    }
+
+    return Err(anyhow!(
```

Comment:
Just to be consistent with the rest of the code, you can remove the `return` here.

Reply:
Thanks, I'll fix it